### PR TITLE
Damage logic + Double Defense in logic

### PR DESCRIPTION
**Pull request by puddydoo**
Based on #1225 but does not include heat/drown timers, just specifies the amount of damage taken as `can_live_dmg()` in all logic.
The function is expanded to take into account starting hearts (but not collectible heart containers) as well as Double Defense, which can potentially be required in quadruple damage mode.
It is in logic that the Goron City grotto can require 4 fairy bottles in OHKO mode, but this condition is not used because there is no real function to count the number of bottles.

### CODE REVIEW

File: World.py [hunk 1]
```diff
@@ -739,6 +750,8 @@ def update_useless_areas(self, spoiler):
                     # to logic
                     dupe_locations = duplicate_item_woth[world_id].get('Bottle', [])
                     max_progressive = 1
+                    # if self.damage_multiplier == 'ohko' and not self.shuffle_grotto_entrances:
+                    #    max_progressive = 4
```
Comment (Zannick) [inline World.py:754]: Please remove.
```suggestion
```

File: data/LogicHelpers.json [hunk 2]
```diff
@@ -61,6 +62,8 @@
             or (for_age == adult and (Bow or Hookshot))
             or (for_age == both and (Slingshot or Boomerang) and (Bow or Hookshot))
             or (for_age == either and (Slingshot or Boomerang or Bow or Hookshot))",
+    # Placeholder for multiple bottle requirements
+    "has_bottles(number)": "False",
```
Comment (Zannick) [inline data/LogicHelpers.json:66]: Just hold off on this for now.

```suggestion
```

**Review by Zannick — CHANGES_REQUESTED** [review]
I may want someone to look over the logic.
Other than that, seems okay, mostly a refactor of the damage helpers and adding DD to be potentially a requirement in quadruple damage.

**Comment by r0bd0g** [comment]
I don't think it's safe to use starting hearts. Could it be possible that you had to previously take damage and there are no heart refills available? 3 hearts is safe because you can always just die to refill to 3, and I don't think there is any case where more than one damage source would need to be accounted for in a row since the last loading zone. Also, if starting hearts can be used for damage sources, you'd expect for consistency that they would also affect how long it's safe to spend in an environmental timer as well, which of course is a whole can of worms.

I'm not sure what's going on with the logic coming in and out of goron city grotto, but I don't trust it. There are a lot of strategies here that you could use to change the amount of hits that you take. It'll be hard to gauge exactly how many hits is fair...

At some point in the past I decided I wouldn't bother with accounting for double defence in logic. We do treat it a bit differently from the other health upgrades, to be fair. I wonder whether full logic accounting for dd, poh's, and hc would be worth it...

**Comment by puddydoo** [comment]
> I don't think it's safe to use starting hearts. Could it be possible that you had to previously take damage and there are no heart refills available? 3 hearts is safe because you can always just die to refill to 3, and I don't think there is any case where more than one damage source would need to be accounted for in a row since the last loading zone. Also, if starting hearts can be used for damage sources, you'd expect for consistency that they would also affect how long it's safe to spend in an environmental timer as well, which of course is a whole can of worms.
> 
> I'm not sure what's going on with the logic coming in and out of goron city grotto, but I don't trust it. There are a lot of strategies here that you could use to change the amount of hits that you take. It'll be hard to gauge exactly how many hits is fair...
> 
> At some point in the past I decided I wouldn't bother with accounting for double defence in logic, considering it just another health upgrade. We do treat it a bit differently from the other health upgrades, to be fair... But also, the double defense still factors in after a death, where you respawn with 3 hearts...

This is why exits don't require taking damage except in certain cases where ER is disabled and there is a guaranteed heart refill or no required damage after that exit. Otherwise it's just locations, and the logic expects you to be able to access each location at full health.

**Comment by r0bd0g** [comment]
I think I can construct a problem scenario on main branch using goron city grotto. On quad damage, you take 3 hits walking across it, damaging you from your starting hearts down to 3 hearts or less. It's dampe's grave and passing through leads to colossus where you find MQ Fire inside. Then you have to run across the lava to get the near boss chest, which takes 3 hits from the lava floor. If heart drops are off I think it could actually be possible that there is no way to refill your health back up to maximum? I really had to reach to come up with this, though. It's a bit easier to think of problem scenarios on Roman fork I guess.

**Comment by puddydoo** [comment]
That particular exit check could probably be made to work differently in that scenario when ER is on, or require more hearts so that it can cover any possible future damage check.

**Comment by cjohnson57** [comment]
I do like the idea of a damage helper that takes a number and does calculations, but some of the changes here are questionable and the branch is far out of date. If you'd like to take another crack at this some time, I would welcome another PR.

**Pull request closed without merging**